File: aoc-py/solutions/day1.py
"""
Day 1: Secret Entrance

https://adventofcode.com/2025/day/1
"""
__all__ = ('Day1',)

from typing import ClassVar

from ..solution import Solution

class Day1(Solution):
    NAME: ClassVar[str] = 'Secret Entrance'

    def part_one(self, inp: str) -> int:
        count = 0
        dial = 50

        for line in inp.splitlines():
            value = int(line[1:])

            if line[0] == 'L':
                value *= -1

            dial += value
            dial %= 100

            if dial == 0:
                count += 1
        return count

    def part_two(self, inp: str) -> int:
        count = 0
        dial = 50

        for line in inp.splitlines():
            value = int(line[1:])

            if line[0] == 'L':
                value *= -1

            # edge case: ex: dial=0 and [L5]
            # should not add to count since `0` was reached last time
            # but -5 // 100 = 1 (not 0)
            if dial == 0 and value < 0:
                count -= 1

            dial += value

            zeroes, dial = divmod(dial, 100)
            count += abs(zeroes)

            # edge case: ex: dial=20 and [L20]
            # dial=0 now but 20 % 100 = 0 and we did pass by `0`
            if dial == 0 and value < 0:
                count += 1
        return count

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 1195
        assert p2 == 6770

File: src/bin/day1.rs
//! Day 1: Secret Entrance
//!
//! <https://adventofcode.com/2025/day/1>
use std::fmt::Display;
use aoc_2025::Solution;

pub struct Day1;

impl Solution for Day1 {
    const NAME: &'static str = "Secret Entrance";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        let mut dial = 50;

        inp
            .to_string()
            .lines()
            .filter(|line| {
                let (op, val) = line.split_at(1);
                let mut val = val.parse::<isize>().unwrap();

                if op == "L" { val *= -1; }

                dial += val;
                dial = dial.rem_euclid(100);

                dial == 0
            })
            .count()
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        let mut dial = 50;

        inp
            .to_string()
            .lines()
            .map(|line| {
                let (op, val) = line.split_at(1);
                let mut val = val.parse::<isize>().unwrap();

                if op == "L" { val *= -1; }

                let mut zeroes = if dial == 0 && val < 0 { -1 } else { 0 };
                dial += val;

                zeroes += dial.div_euclid(100).abs();
                dial = dial.rem_euclid(100);

                if dial == 0 && val < 0 { zeroes += 1; }

                zeroes
            })
            .sum::<isize>().cast_unsigned()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 1195);
        assert_eq!(p2, 6770);
    }
}

fn main() {
    aoc_2025::run_day(1, &Day1);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}